Add strike through animation to homework

nilsreichardt opened the pull request:
https://github.com/user-attachments/assets/c96aaf69-3431-4548-b871-5f37ff89f2d1

ToDos:

- [ ] ~Add option to hide strike through on completed list?~
- [ ] ~Strike through for homework with submissions~
- [x] Sometimes the homework doesn't disappear from the completed list
- [ ] ~Update tests?~
- [ ] ~Is there an option to display the animation without delaying the database call~ (semi fixed)
- [x] Support dark mode

nilsreichardt commented:
@Jonas-Sander WDYT?

Jonas-Sander commented:
I think:
* the animation speed at the start should be faster. In the video it ramps up gradually. A "quicker" curve should fit better I guess.
* I think on the completed page they shouldn't be crossed through at all since it makes it hard to read.
* Not sure if the abbreviation should get crossed through (or maybe the strike should cover the whole circle. not just the text)

Sharezone-Admin commented:
⬇️ Builds generated by [Codemagic](https://codemagic.io) for commit 59f44b9bf0dd4e57804ff389c4f34dac550f7b2c ⬇️

_Note_: Only Sharezone team members are able to install the iOS app.

| iOS | | Android |
|:-:|:-:|:-:|
| ![iOS QR code](https://app-preview-qr.nils.re/?size=150x150&data=https%3A%2F%2Fapi.codemagic.io%2Fartifacts%2F.eJwVwcmOgjAAANB_8W4CiB08eKjIJrUyVNkuBIGySgpVC3z9ZN7bVPDfaTh3eUPotVQQgdXWyArwE5zGEOGEH7hpgTYXPrOq6fKq1_BmlA6UUviysF3ZATX075K5C7cShCN5Pqgawpw_-14pnky-LF_5Pfkw-xTsQaPU0HAhS7g7K0ZGhJ_q92u3g2pHAm2_XYeIAWh7da_j6WbOCFueghTBB7DsSt8JtL59uW-D0zGRHEru6FAriZfv4yCnDhFiyRFpffkz9gUbmNjVZxqmiAm21lhuZwpKNN1Ut7JBswzjL1BzwSM9i523SFY9araPELimGcenaXh436Y6Hjd_f6djXg.hV_BhWmZT8NbSr2p44Pm9UAHX5I&platform=ios&groupId=6f933136) <br /> [Download link](https://api.codemagic.io/artifacts/.eJwVwcmOgjAAANB_8W4CiB08eKjIJrUyVNkuBIGySgpVC3z9ZN7bVPDfaTh3eUPotVQQgdXWyArwE5zGEOGEH7hpgTYXPrOq6fKq1_BmlA6UUviysF3ZATX075K5C7cShCN5Pqgawpw_-14pnky-LF_5Pfkw-xTsQaPU0HAhS7g7K0ZGhJ_q92u3g2pHAm2_XYeIAWh7da_j6WbOCFueghTBB7DsSt8JtL59uW-D0zGRHEru6FAriZfv4yCnDhFiyRFpffkz9gUbmNjVZxqmiAm21lhuZwpKNN1Ut7JBswzjL1BzwSM9i523SFY9araPELimGcenaXh436Y6Hjd_f6djXg.hV_BhWmZT8NbSr2p44Pm9UAHX5I) | | ![Android QR code](https://app-preview-qr.nils.re/?size=150x150&data=https%3A%2F%2Fapi.codemagic.io%2Fartifacts%2F.eJwVwcmSQ0AAANB_yV1VYhsOOaBJ0JaEijaXVFuatrQgQXz91Lx3qLR_OgNt3tg4LyVIhMQYhEdQlJHqTThr-KJ7iXA9uhEaXB6RZ4ss_82m27tKi-9njTp6p8Nv8iBkG-qbOubhCu0c9GWtTePI-s_-zPfO8RWUpwgu2TW2cEbMOYux4lu1S1VX1k6-atjBqsgKS6UN9F8-B4jBhR9nnaPXmcakF9rNZkYaK7wGwCRnzhqswys1LnvdNSa9120sihk6-Rd9ok5iIT-0BS602cPDlXmNblgat6J-Wr04e3MS4LjhX3StFuiZIIDc5JHyHi7yjokz_oS6tx-56nw-_AFYZWQ7.bj35595yKCYtw0v_nDbJnlf1bLY&platform=android&groupId=6f933136) <br /> [Download link](https://api.codemagic.io/artifacts/.eJwVwcmSQ0AAANB_yV1VYhsOOaBJ0JaEijaXVFuatrQgQXz91Lx3qLR_OgNt3tg4LyVIhMQYhEdQlJHqTThr-KJ7iXA9uhEaXB6RZ4ss_82m27tKi-9njTp6p8Nv8iBkG-qbOubhCu0c9GWtTePI-s_-zPfO8RWUpwgu2TW2cEbMOYux4lu1S1VX1k6-atjBqsgKS6UN9F8-B4jBhR9nnaPXmcakF9rNZkYaK7wGwCRnzhqswys1LnvdNSa9120sihk6-Rd9ok5iIT-0BS602cPDlXmNblgat6J-Wr04e3MS4LjhX3StFuiZIIDc5JHyHi7yjokz_oS6tx-56nw-_AFYZWQ7.bj35595yKCYtw0v_nDbJnlf1bLY) |

<sub>(expires Thu, 14 Nov 2024 10:42:40 GMT)<br/>Codemagic build: [670e43e9e8144e9023592391](https://codemagic.io/app/629c82ea463af7ff553fc7a5/build/670e43e9e8144e9023592391); comment generated with [Codemagic App Preview](https://github.com/nilsreichardt/codemagic-app-preview)</sub>

<!-- Codemagic App Preview; appName: default -->

nilsreichardt commented on app/lib/homework/shared/homework_tile_template.dart:
```diff
@@ -70,3 +74,166 @@ class HomeworkTileTemplate extends StatelessWidget {
+              overflow: TextOverflow.ellipsis,
+              // We can only show one line of text in the title. Showing more
+              // lines would make the strike through look weird.
+              maxLines: 1,
```
We need here to make a trade-off and reduce the `maxLines` from 2 to 1. I think it's okay since you normally know what the homework was when reading the first line. Otherwise, you just press on the homework and you get the full title.

nilsreichardt commented:
> the animation speed at the start should be faster. In the video it ramps up gradually. A "quicker" curve should fit better I guess.

Changed it to `Curve.ease`, should be better now 👍 

> I think on the completed page they shouldn't be crossed through at all since it makes it hard to read.

Removed it - I was also unsure about show it there.

> Not sure if the abbreviation should get crossed through (or maybe the strike should cover the whole circle. not just the text)

I kept it because striking the whole circle looked weird and not striking it also looked weird because everything gets struck expect the abbreviation on the left.

Jonas-Sander reviewed (approved):
I'd personally would like the animation to be a bit quicker still, but I'm totally okay with the current state as well. LGTM!
